core/src/main/java/org/apache/iceberg/avro/ValueWriters.java

```diff
@@ -45,68 +55,72 @@ private ValueWriters() {
-    return ByteToIntegerWriter.INSTANCE;
+  public static ValueWriter<Byte> tinyints(int id) {
+    return new ByteToIntegerWriter(id);
   }
 
-  public static ValueWriter<Short> shorts() {
-    return ShortToIntegerWriter.INSTANCE;
+  public static ValueWriter<Short> shorts(int id) {
+    return new ShortToIntegerWriter(id);
   }
 
-  public static ValueWriter<Integer> ints() {
-    return IntegerWriter.INSTANCE;
+  public static ValueWriter<Integer> ints(int id) {
+    return new IntegerWriter(id);
   }
 
-  public static ValueWriter<Long> longs() {
-    return LongWriter.INSTANCE;
+  public static ValueWriter<Long> longs(int id) {
+    return new LongWriter(id);
   }
 
-  public static ValueWriter<Float> floats() {
-    return FloatWriter.INSTANCE;
+  public static ValueWriter<Float> floats(int id) {
+    return new FloatWriter(id);
   }
 
-  public static ValueWriter<Double> doubles() {
-    return DoubleWriter.INSTANCE;
+  public static ValueWriter<Double> doubles(int id) {
+    return new DoubleWriter(id);
   }
 
-  public static ValueWriter<Object> strings() {
-    return StringWriter.INSTANCE;
+  public static ValueWriter<CharSequence> strings(int id) {
+    return new StringWriter(id);
   }
 
-  public static ValueWriter<Utf8> utf8s() {
-    return Utf8Writer.INSTANCE;
+  public static ValueWriter<Utf8> utf8s(int id) {
+    return new Utf8Writer(id);
   }
 
-  public static ValueWriter<UUID> uuids() {
-    return UUIDWriter.INSTANCE;
+  public static ValueWriter<UUID> uuids(int id) {
+    return new UUIDWriter(id);
   }
 
-  public static ValueWriter<byte[]> fixed(int length) {
-    return new FixedWriter(length);
+  public static ValueWriter<byte[]> fixed(int id, int length) {
+    return new FixedWriter(id, length);
   }
 
-  public static ValueWriter<GenericData.Fixed> genericFixed(int length) {
-    return new GenericFixedWriter(length);
+  public static ValueWriter<GenericData.Fixed> genericFixed(int id, int length) {
+    return new GenericFixedWriter(id, length);
   }
 
-  public static ValueWriter<byte[]> bytes() {
-    return BytesWriter.INSTANCE;
+  public static ValueWriter<byte[]> bytes(int id) {
+    return new BytesWriter(id);
   }
 
-  public static ValueWriter<ByteBuffer> byteBuffers() {
-    return ByteBufferWriter.INSTANCE;
+  public static ValueWriter<ByteBuffer> byteBuffers(int id) {
+    return new ByteBufferWriter(id);
   }
 
-  public static ValueWriter<BigDecimal> decimal(int precision, int scale) {
-    return new DecimalWriter(precision, scale);
+  public static ValueWriter<BigDecimal> decimal(int id, int precision, int scale) {
+    return new DecimalWriter(id, precision, scale);
   }
 
-  public static <T> ValueWriter<T> option(int nullIndex, ValueWriter<T> writer) {
-    return new OptionWriter<>(nullIndex, writer);
+  public static <T> ValueWriter<T> option(int nullIndex, ValueWriter<T> writer, Schema.Type type) {
+    if (AvroSchemaUtil.supportsMetrics(type)) {
```

Comment:
Rather than introducing `supportsMetrics`, why not just check the value writer to see whether it is a `MetricsAwareWriter`? I know that not all of the writers extend that class, but you could either introduce a `MetricsWriter` interface to signal that the inner writer supports metrics that all of the implementations extend, or maybe you could alter the hierarchy a little so that `StoredAsIntWriter` and `StoredAsLongWriter` actually do extend `MetricsAwareWriter`. Then you wouldn't need changes to `AvroSchemaUtil` or so many changes to option handling.